vscode/src/circuit.ts

```diff
@@ -0,0 +1,134 @@
+  }
+
+  sendMessageToPanel("circuit", true, undefined);
+
+  let timeout = false;
+
+  // Start the worker, run the code, and send the results to the webview
+  const worker = getCompilerWorker(compilerWorkerScriptPath);
+  const compilerTimeout = setTimeout(() => {
+    timeout = true;
+    sendTelemetryEvent(EventType.CircuitEnd, {
+      associationId,
+      reason: "timeout",
+      flowStatus: UserFlowStatus.Aborted,
+    });
+    log.info("terminating circuit worker due to timeout");
+    worker.terminate();
+  }, compilerRunTimeoutMs);
+  let title;
+  let subtitle;
+  const targetProfile = getTarget();
+  const sources = await loadProject(editor.document.uri);
+  if (operation) {
+    title = `${operation.operation} with ${operation.totalNumQubits} input qubits`;
+    subtitle = `${getTargetFriendlyName(targetProfile)} `;
+  } else {
+    title = basename(editor.document.uri.path) || "Circuit";
+    subtitle = `${getTargetFriendlyName(targetProfile)}`;
+  }
+
+  try {
+    sendTelemetryEvent(EventType.CircuitStart, { associationId }, {});
+    const circuit = await worker.getCircuit(sources, targetProfile, operation);
+    clearTimeout(compilerTimeout);
+
+    const message = {
+      command: "circuit",
+      circuit,
+      title,
+      subtitle,
+    };
+    sendMessageToPanel("circuit", false, message);
+
+    sendTelemetryEvent(EventType.CircuitEnd, {
+      associationId,
+      flowStatus: UserFlowStatus.Succeeded,
+    });
+  } catch (e: any) {
+    log.error("Circuit error. ", e.toString());
+    clearTimeout(compilerTimeout);
+
+    const errors: [string, VSDiagnostic][] =
+      typeof e === "string" ? JSON.parse(e) : undefined;
+    let errorHtml = "There was an error generating the circuit.";
+    if (errors) {
+      errorHtml = errorsToHtml(errors);
+    }
+
+    if (!timeout) {
+      sendTelemetryEvent(EventType.CircuitEnd, {
+        associationId,
+        reason: errors && errors[0] ? errors[0][1].code : undefined,
+        flowStatus: UserFlowStatus.Failed,
+      });
+    }
+
+    const message = {
+      command: "circuit",
+      title,
+      subtitle,
+      errorHtml,
+    };
+    sendMessageToPanel("circuit", false, message);
+  } finally {
+    log.info("terminating circuit worker");
+    worker.terminate();
+  }
+}
+
+function errorsToHtml(errors: [string, VSDiagnostic][]): string {
+  let errorHtml = "";
+  for (const error of errors) {
+    const uri = Uri.parse(error[0]);
+    const openCommandUri = Uri.parse(
+      `command:vscode.open?${encodeURIComponent(JSON.stringify([uri]))}`,
+    );
+
+    const fsPath = escapeHtml(uri.fsPath);
+    const errorText = escapeHtml(
+      `:${error[1].range.start.line}:${error[1].range.start.character}: ${error[1].message} (${error[1].code})`,
+    );
+    errorHtml += `<a href="${openCommandUri}">${fsPath}</a>${errorText}<br/>`;
+  }
+  return errorHtml;
+}
+
+function escapeHtml(unsafe: string): string {
```

Comment:
Is encodeUriComponent not suitable here? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent (Or `encodeURI` - I can never remember which to use when)

Reply:
(Actually, I guess that' for URLs, not for escaping HTML characters. Ignore me)

Reply:
This gets me every time. I'm outraged there's no built-in HTML escape function in JavaScript...